base/src/protocol_state_proof/mod.rs

```diff
@@ -0,0 +1,144 @@
+
+pub mod bulletproof_challenges;
+pub use bulletproof_challenges::{
+    BulletproofChallenge, BulletproofChallengeTuple17, BulletproofChallengeTuple18,
+    BulletproofChallenges, BulletproofPreChallenge, ProofStateBulletproofChallenges,
+    ScalarChallengeVector2,
+};
+
+pub mod field_and_curve_elements;
+pub use field_and_curve_elements::{
+    ECPoint, ECPointVec, FieldElement, FieldElementVec, FiniteECPoint, FiniteECPointPairVec,
+    FiniteECPointVec,
+};
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+#[wire_type(recurse = 4)]
+pub struct ProtocolStateProof {
+    pub statement: ProofStatement,
+    pub prev_evals: PrevEvals,
+    pub prev_x_hat: PrevXHat,
+    pub proof: Proof,
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+#[wire_type(recurse = 2)]
+pub struct ProofStatement {
+    pub proof_state: ProofState,
+    pub pass_through: PairingBased,
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+pub struct ProofState {
+    pub deferred_values: ProofStateDeferredValues,
+    pub sponge_digest_before_evaluations: SpongeDigestBeforeEvaluations,
+    pub me_only: ProofStatePairingBased,
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+pub struct ProofStateDeferredValues {
+    pub plonk: Plonk,
+    pub combined_inner_product: ShiftedValue,
+    pub b: ShiftedValue,
+    pub xi: BulletproofPreChallenge,
+    pub bulletproof_challenges: BulletproofChallengeTuple18,
+    pub which_branch: Char,
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+pub struct Plonk {
+    pub alpha: BulletproofPreChallenge,
+    pub beta: ScalarChallengeVector2,
+    pub gamma: ScalarChallengeVector2,
+    pub zeta: BulletproofPreChallenge,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+#[non_exhaustive]
+pub enum ShiftedValue {
+    ShiftedValue(BigInt256),
+}
+
+impl Default for ShiftedValue {
+    fn default() -> Self {
+        Self::ShiftedValue(BigInt256::default())
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug, WireType)]
+#[serde(from = "<Self as WireType>::WireType")]
+#[serde(into = "<Self as WireType>::WireType")]
+#[wire_type(recurse = 2)]
+pub struct SpongeDigestBeforeEvaluations((Hex64, Hex64, Hex64, Hex64, ()));
```

Comment:
is the empty unit needed for a byte in binprot or is it something from arkworks?

Reply:
It is for bin-prot. They use null byte terminated arrays and this is a nice trick for deserializing those with serde